channels/offline-channel-js/src/OfflineBatchHandler.ts

```diff
@@ -0,0 +1,387 @@
+
+
+            _initDefaults();
+
+            _self.initialize = (providerContext: ILocalStorageProviderContext) => {
+                // TODO: handle provider change
+                try {
+                    if (!_isInitialized) {
+                        _logger = logger || providerContext.itemCtx.diagLog();
+                        _itemCtx = providerContext.itemCtx;
+                        _provider = _initProvider(providerContext);
+                        let storeCfg = providerContext.storageConfig || {};
+                        let retry = storeCfg.maxRetry;
+                        _maxRetryCnt = !isNullOrUndefined(retry)? retry : 2;
+                        _retryCodes = (storeCfg.senderCfg || {}).retryCodes;
+                        if (_provider) {
+                            _isInitialized = true;
+                        }
+                    }
+                   
+                } catch(e) {
+                    _isInitialized = false;
+                    
+                }
+                return _isInitialized;
+                
+            }
+
+            _self["_getDbgPlgTargets"] = () => {
+                return [_provider, _isInitialized, _unloadProvider];
+            };
+            
+
+            _self.storeBatch = (batch: IPayloadData, cb?: OfflineBatchStoreCallback, sync?: boolean) => {
+                
+                let provider = _provider;
+                if(!!sync) {
+                    provider = _unloadProvider;
+                }
+                return createAsyncPromise((resolve, reject) => {
+                    if (!provider) {
+                        reject(new Error(NoProviderErrMsg))
+                    }
+                    let evt = _getOfflineEvt(batch);
+                    return doAwaitResponse(provider.addEvent(evt.id, evt, _itemCtx), (response: AwaitResponse<IStorageTelemetryItem>) => {
+                        try {
+                            let evt = response.value || response.reason || [];
+                            let state = eBatchStoreStatus.Failure;
+                            if (!response.rejected) {
+                                state = eBatchStoreStatus.Success;
+                            }
+                            let res = {state: state, item: evt};
+                            cb && cb(res);
+                            resolve(res);
+                        } catch(e) {
+                            reject(e);
+                        }
+                    });
+                });
+
+            }
+
+            _self.sendNextBatch = (cb?: OfflineBatchCallback, sync?: boolean, xhrOverride?: IXHROverride, cnt: number = 1) => {
+                return createAsyncPromise((resolve, reject) => {
+                    let provider = _provider;
+                    try {
+                       
+                        if (_unloadProvider) {
+                            doAwaitResponse(_unloadProvider.getNextBatch(), (res) => {
```

Comment:
Maybe this should return here and after the promise has resolve / rejected it either returns the "unload" batch, or simply calls the code below (which would need to be extracted out to it's own function so that it can be called below or from within this async handler.